### PR TITLE
Add support for sending and receiving stereo audio stream

### PR DESCRIPTION
**Pull request by pracheth**
**Issue #:** 

The additions from the reviews in private repo are the new drop down options added in microphone and content share dropdowns to make testing easier. 

**Description of changes:**
This commit adds support to send and receive stereo audio on JS SDK

**Testing:**
* Newly added stereo integration tests succeed using test demo URL
* QA testing has been performed with the code changes from private repo
* Also added unit tests

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*
1. Build and deploy a demo application using the code changes OR contact for test URL
2. On the join screen, check the box the says `Set fullband music (stereo) quality`
3. Join the meeting using step 2 from two browsers
4. Click microphone drop down and select `Prerecorded speech (stereo)`  from one of the browsers
5. On the other browser, you should be able to hear the audio for left channel in the left earphone and audio for right channel in the right earphone. Recommend using a wired headset. 

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
Y

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
Y

3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
N

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/devicecontroller/DefaultDeviceController.ts [hunk 1]
```diff
@@ -1074,6 +1074,25 @@ export default class DefaultDeviceController
     return outputNode.stream;
   }
 
+  static synthesizeStereoTones(toneHzLeft: number, toneHzRight: number): MediaStream {
```
Comment (ltrung) [inline src/devicecontroller/DefaultDeviceController.ts:1077]: Can we add a doc on what this method do?

Comment (pracheth) [reply src/devicecontroller/DefaultDeviceController.ts:1077]: Sure. Where would be a good place to add that?

Comment (ltrung) [reply src/devicecontroller/DefaultDeviceController.ts:1077]: You can just add it here in this method. For instance methods, we document in the interface level but for static we don't have one standard place yet. @simmkyu

File: src/audioprofile/AudioProfile.ts [hunk 2]
```diff
@@ -32,4 +32,18 @@ export default class AudioProfile {
...
+   */
+  static fullbandMusicStereo(): AudioProfile {
+    return new AudioProfile(128000);
+  }
+
+  /**
+   * Returns true if audio profile is set to stereo mode.
+   */
+  isStereo(): boolean {
+    return this.audioBitrateBps === AudioProfile.fullbandMusicStereo().audioBitrateBps;
```
Comment (ltrung) [inline src/audioprofile/AudioProfile.ts:47]: nit: Seem a bit roundabout to just compare it to 12800 right?

File: src/sdp/DefaultSDP.ts [hunk 3]
```diff
@@ -218,41 +218,75 @@ export default class DefaultSDP implements SDP {
...
-      if (line.startsWith('m=video')) {
-        lookingForOpusRtpMap = false;
-        fmtpAttribute = null;
       }
       if (lookingForOpusRtpMap) {
         const match = opusRtpMapRegex.exec(line);
         if (match !== null) {
           fmtpAttribute = `a=fmtp:${match[1]} `;
-          lookingForOpusRtpMap = false;
+          break;
```
Comment (ltrung) [inline src/sdp/DefaultSDP.ts:250]: Do we change the logic here? It seems previously we loop through every line in sdp but now we quit after finding the opus line. Does that mean there can be only one opus line?

File: src/sdp/DefaultSDP.ts [hunk 4]
```diff
@@ -218,41 +218,75 @@ export default class DefaultSDP implements SDP {
...
+    const srcLines: string[] = this.lines();
+    const fmtpAttribute = DefaultSDP.findOpusFmtpAttribute(srcLines);
+    const dstLines = DefaultSDP.updateOpusFmtpAttribute(srcLines, fmtpAttribute, [
+      'stereo=1',
+      'sprop-stereo=1',
+    ]);
+    return DefaultSDP.linesToSDP(dstLines);
+  }
+
+  static findOpusFmtpAttribute(sdpLines: string[]): string | null {
```
Comment (ltrung) [inline src/sdp/DefaultSDP.ts:238]: SDP code is very hard to understand for me and other people in JS SDK. Can we put documentation on  both these static method and maybe more comment inside the method on which loop does what.

File: README.md [hunk 5]
```diff
@@ -1159,9 +1159,20 @@ Use the following setting to optimize the audio bitrate of content share audio f
 meetingSession.audioVideo.setContentAudioProfile(AudioProfile.fullbandMusicMono());
 ```
 
+**Use case 34.** Set the audio quality of content share audio to optimize for music with stereo channel:
```
Comment (ltrung) [inline README.md:1162]: Is there any plan to write a more in-depth guide on audio? For now, can we add when they want to enable this?
The use case here is for content audio but for testing we just test audio so seem a bit weird.

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add `supportDownlinkBandwidthEstimation` API to check whether browsers support downlink bandwidth estimation 
   which requires for priority based downlink policy to work.
 - Add `keepLastFrameWhenPaused` in `DefaultVideoTile` as an option to keep last frame when pausing a video tile.
+- Add support to send and receive stereo audio
```
Comment (ltrung) [inline CHANGELOG.md:13]: Can we add a bit about the use case that builders would want to use this here?

File: src/devicecontroller/DefaultDeviceController.ts (hunk 1 above)
Comment (ltrung) [reply src/devicecontroller/DefaultDeviceController.ts:1077]: Also is this API only used for testing? If so should it be here or should it just be in the demo app.

Comment (pracheth) [reply src/devicecontroller/DefaultDeviceController.ts]: Added comment in latest push and moved into `meetingV2.ts` as its only used for testing

File: src/sdp/DefaultSDP.ts (hunk 3 above)
Comment (pracheth) [reply src/sdp/DefaultSDP.ts]: Ah good catch. There can actually be multiple opus lines. I updated the code and unit test to account for this. Also added comments.

File: src/sdp/DefaultSDP.ts (hunk 4 above)
Comment (pracheth) [reply src/sdp/DefaultSDP.ts]: Added some comments

File: README.md (hunk 5 above)
Comment (pracheth) [reply README.md]: Updated this section a bit.

File: CHANGELOG.md (hunk 6 above)
Comment (pracheth) [reply CHANGELOG.md]: Updated changelog. Also added demo updates.

File: src/sdp/DefaultSDP.ts [hunk 7]
```diff
@@ -218,41 +218,95 @@ export default class DefaultSDP implements SDP {
...
+    ]);
+    return DefaultSDP.linesToSDP(dstLines);
+  }
+
+  // Here we loop through each line in the SDP
+  // and construct an array containing the fmtp
+  // attribute for all the audio m lines that use
+  // the opus codec. If it doesn't use opus codec
+  // we add null to the array
+  static findOpusFmtpAttributes(sdpLines: string[]): string[] | null {
```
Comment (ltrung) [inline src/sdp/DefaultSDP.ts:243]: nit: Could we use the same format use everywhere else?
/**
*
**/

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 8]
```diff
@@ -2738,26 +2765,16 @@ export class DemoMeetingApp
...
-        mediaElementSource.connect(streamDestination);
-        return streamDestination.stream;
-      } catch (e) {
-        this.log(`Error fetching audio from ${audioPath}: ${e}`);
-        return null;
-      }
+      return this.streamAudioFile('audio_file');
+    }
+
+    if (value == 'Prerecorded Speech (Stereo)') {
```
[the remark this answers is not in the document]
Comment (pracheth) [reply demos/browser/app/meetingV2/meetingV2.ts]: Updated

File: demos/browser/app/meetingV2/meetingV2.ts [hunk 9]
```diff
@@ -98,6 +98,8 @@ let SHOULD_DIE_ON_FATALS = (() => {
   return fatalYes || (isLocal && !fatalNo);
 })();
 
+let USING_STEREO_MUSIC_AUDIO_PROFILE = false;
```
[the remark this answers is not in the document]
Comment (pracheth) [reply demos/browser/app/meetingV2/meetingV2.ts]: Updated

File: CHANGELOG.md [hunk 10]
```diff
@@ -10,6 +10,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- [Demo] Add new dropdown items in microphone dropdown menu to test sending stereo audio as main audio input
+- [Demo] Add new dropdown items in content share dropdown menu to test sending stereo audio as content
+
+### Removed
+
+### Fixed
+
+### Changed
+
+## [2.24.0] - 2021-12-07
```
[the remark this answers is not in the document]
Comment (pracheth) [reply CHANGELOG.md]: Ah this got messed up during a rebase I did earlier. Fixed in latest push

File: src/sdp/DefaultSDP.ts (hunk 7 above)
Comment (pracheth) [reply src/sdp/DefaultSDP.ts:247]: Updated

File: src/audioprofile/AudioProfile.ts (hunk 2 above)
Comment (pracheth) [reply src/audioprofile/AudioProfile.ts]: Updated

File: CHANGELOG.md [hunk 11]
```diff
@@ -13,6 +13,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add error name for custom device controller error.
 - Added pagination option to meeting demo when priority downlink policy is used.
 - Add `ApplicationMetadata` to enable builders to send their application name or version to the Amazon Chime backend. This is an opt-in addition.
+- Add a new `AudioProfile` called `fullbandMusicStereo` which can be passed into `setAudioProfile` to support sending and recieving stereo audio through main audio input and output. This can also be passed into `setContentAudioProfile` to support sending stereo audio as content
+- [Demo] Add new checbox on join screen to select new `fullbandMusicStereo` audio profile
```
Comment (ltrung) [inline CHANGELOG.md:17]: nit: You don't need entries for demo. CHANGELOG is only for library change.

Comment (pracheth) [reply CHANGELOG.md:17]: I saw a few [Demo] entries in the changelog hence added them. Example, https://github.com/aws/amazon-chime-sdk-js/blob/master/CHANGELOG.md?plain=1#L237

**Review by achronop — APPROVED** [review]

File: CHANGELOG.md (hunk 11 above)
Comment (ltrung) [reply CHANGELOG.md:17]: Yeah they just do it wrong. Anyway, just to let  you know no need to change unless you need to fix something else :)

**Review by ltrung — APPROVED** [review]

**Pull request merged**